File: Solutions/c/ex3-ring/ring.c

/******************************************************************************
 *
 * Accumulate the sum of the ranks by sending a message around a
 * ring.
 *
 *  We use a synchronous non-blocking send mpi_issend() and a non-
 *  blocking recieve mpi_irecv(), along with two wait statements.
 *
 *****************************************************************************/

#include <mpi.h>
#include <stdio.h>

int main (int argc, char* argv[]){

  int i, rank, size;
  int my_prev, my_next;
  int sum;

  /*  Message information  */
  int sendbuf, recvbuf;
  MPI_Request req[2];
  int my_tag = 9999;
  MPI_Status status[2];

  /* loop counter */
  int n;

  MPI_Init(&argc,&argv);

  MPI_Comm_rank(MPI_COMM_WORLD, &rank);
  MPI_Comm_size(MPI_COMM_WORLD, &size);

  /* Work out neighbouring ranks */
  my_prev = rank - 1;
  
  if (my_prev < 0) 
  	my_prev = my_prev + size;
  
  my_next = (rank + 1) % size;

  /* Initialise the sum and the message for all ranks */
  sum = 0;
  sendbuf = rank;
        
  for(i = 0; i < size; i++) {
  	MPI_Irecv(&recvbuf, 1, MPI_INT, my_prev, my_tag, MPI_COMM_WORLD, req);
        MPI_Issend(&sendbuf, 1, MPI_INT, my_next, my_tag, MPI_COMM_WORLD, req+1);
           
        /* When the receive has completed, we can use the contents of recvbuf */
        MPI_Wait(req, status);
        sum = sum + recvbuf;

        /* When the send has completed, we can safely re-use the send buffer */
        MPI_Wait(req, status);
        sendbuf = recvbuf;
    
  }

  /* Display the result on all ranks, along with the correct answer */

  printf("Rank %2d has sum of ranks %d; Answer = %d\n",rank, sum, (size-1)*size/2);

  MPI_Finalize();
	
  return 0;

} 

File: Exercises/fortran/ex3-ring/ring.f90
program ring

  use mpi
  use ifport ! Intel fortran specific

  implicit none

  integer :: ierr, rrequest, srequest
  integer :: rank, size, left, right
  integer :: sum

  integer :: recbuf, sendbuf

  integer, parameter :: tag = 99

  integer :: status(MPI_STATUS_SIZE)


  call mpi_init(ierr)

  call mpi_comm_rank(MPI_COMM_WORLD, rank, ierr)
  call mpi_comm_size(MPI_COMM_WORLD, size, ierr)

  ! Work out neighbouring ranks
  left = mod((rank-1)+size,size)
  right = mod(rank+1,size)

  print *,'Rank = ', rank, ' left neigbour: ', left, ' right neighbour: ',right

  ! Initialise the sum
  sum = 0

  ! Initialise recbuf to a number that cannot be the same as a rank
  recbuf = -1
  sendbuf = rank

  do while (recbuf /= rank) 

    ! Send message to the right
    call mpi_isend(sendbuf, 1, MPI_INTEGER, right, tag, MPI_COMM_WORLD, srequest, ierr)

    print *,'Rank ',rank,' sent ',sendbuf,' to ',right

    ! Receive message from the left
    call mpi_irecv(recbuf, 1, MPI_INTEGER, left, tag, MPI_COMM_WORLD, rrequest, ierr)

    call mpi_wait(rrequest, status, ierr)

    ! Sleep a variable number of seconds based on number just sent 
    print *,'Rank ',rank,' is sleeping ',sendbuf,' seconds'
    call sleep(sendbuf)
    print *,'Rank ',rank,' received ',recbuf,' from ',left

    sum = sum + recbuf

    call mpi_wait(srequest, status, ierr)

    sendbuf = recbuf

  end do

  ! Display the result on all ranks, along with the correct answer
  write (unit = *, fmt = '(a,i3,a,i4,a,i4)') &
       "Rank ", rank, " has sum of ranks ", sum, " Answer ", (size - 1)*size/2

  call mpi_finalize(ierr)


end program ring
